src/main/java/com/google/devtools/build/lib/analysis/platform/ConstraintCollection.java

```diff
@@ -32,46 +33,82 @@
 import com.google.devtools.build.lib.syntax.Printer;
 import com.google.devtools.build.lib.syntax.SkylarkList;
 import com.google.devtools.build.lib.util.Fingerprint;
+import java.util.List;
+import java.util.Map;
 import java.util.function.Function;
 import javax.annotation.Nullable;
 
 /** A collection of constraint values. */
 @Immutable
 @AutoCodec
-public final class ConstraintCollection
+@AutoValue
+public abstract class ConstraintCollection
     implements ConstraintCollectionApi<ConstraintSettingInfo, ConstraintValueInfo> {
-  @Nullable private final ConstraintCollection parent;
-  private final ImmutableMap<ConstraintSettingInfo, ConstraintValueInfo> constraints;
 
-  /** Creates a new {@link ConstraintCollection} with the given constraint values. */
-  public ConstraintCollection(ImmutableList<ConstraintValueInfo> constraints) {
-    this(null, constraints);
+  /** A builder class to help create instances of {@link ConstraintCollection}. */
+  public final static class Builder {
+    @Nullable private ConstraintCollection parent;
+    private ImmutableList.Builder<ConstraintValueInfo> constraintValues = new ImmutableList.Builder<>();
+
+    private Builder() {}
+
+    /** Sets the parent {@link ConstraintCollection} of this instance. */
+    public Builder parent(@Nullable ConstraintCollection parent) {
+      this.parent = parent;
+      return this;
+    }
+
+    /** Adds the given constraints to the current collection. */
+    public Builder addConstraints(Map<ConstraintSettingInfo, ConstraintValueInfo> constraints) {
+      return addConstraints(constraints.values());
+    }
+
+    /** Adds the given constraints to the current collection. */
+    public Builder addConstraints(ConstraintValueInfo... constraints) {
+      return addConstraints(ImmutableList.copyOf(constraints));
+    }
+
+    /** Adds the given constraints to the current collection. */
+    public Builder addConstraints(Iterable<ConstraintValueInfo> constraints) {
+      constraintValues.addAll(constraints);
+      return this;
+    }
+
+    /** Returns the completed {@link ConstraintCollection} instance. */
+    public ConstraintCollection build() {
+      ImmutableList<ConstraintValueInfo> constraintValues = this.constraintValues.build();
+      // TODO: validate
```

Comment:
If you want to defer this, maybe use a `TODO(<bug>)` comment?

Reply:
Oh, it's fixed in (counts) two more PRs for this cycle, but I should address that.